Add Events for Policy Service APIs

adnanhemani opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

MonkeyCanCode committed:
2334 (#2427)

XN137 committed:
Fix TableIdentifier in TaskFileIOSupplier (#2304)

we cant just convert a `TaskEntity` to a `IcebergTableLikeEntity` as the
`getTableIdentifier()` method will not return a correct value by using
the name of the task and its parent namespace (which is empty?).

task handlers instead need to pass in the `TableIdentifier` that they
already inferred via `TaskEntity.readData`.

HonahX committed:
Fix NPE in CreateCatalog (#2435)

MonkeyCanCode committed:
Doc fix: Access control page update (#2424)

* 2418

* 2418

renovate-bot committed:
fix(deps): update dependency software.amazon.awssdk:bom to v2.32.29 (#2443)

XN137 committed:
Optimize PolicyCatalog.listPolicies (#2370)

this is a follow-up to https://github.com/apache/polaris/pull/2290

the optimization is to use `listEntities` instead of `loadEntities` when
there is `policyType` filter to apply

XN137 committed:
Add PolarisDiagnostics field to BaseMetaStoreManager (#2381)

* Add PolarisDiagnostics field to BaseMetaStoreManager

the ultimate goal is removing the `PolarisCallContext` parameter from every
`PolarisMetaStoreManager` interface method, so we make steps towards
reducing its usage first.

dimas-b committed:
Add feature flag to disallow custom S3 endpoints (#2442)

* Add new realm-level flag: `ALLOW_SETTING_S3_ENDPOINTS` (default: true)

* Enforce in `PolarisServiceImpl.validateStorageConfig()`

Fixes #2436

adutra committed:
Deprecate ActiveRolesProvider for removal (#2404)

MonkeyCanCode committed:
Client: fix openapi verbose output, remove doc generate, and skip test generations (#2439)

* Fix various issue in client code generation

* Use logger instead of print

* Add back exclude on __pycache__ as CI is not via Makefile

* Add back exclude on __pycache__ as CI is not via Makefile

fivetran-kostaszoumpatianos committed:
Add user principal tag in metrics (#2445)

* Added API change to enable tag

* Added test

* Added production readiness check

renovate-bot committed:
fix(deps): update dependency io.opentelemetry.semconv:opentelemetry-semconv to v1.36.0 (#2454)

renovate-bot committed:
fix(deps): update dependency com.google.cloud:google-cloud-storage-bom to v2.56.0 (#2447)

renovate-bot committed:
fix(deps): update dependency gradle.plugin.org.jetbrains.gradle.plugin.idea-ext:gradle-idea-ext to v1.3 (#2428)

adnanhemani committed:
Add Events for Policy Service APIs

snazy committed:
Build: Make jandex dependency used for index generation managed (#2431)

Also allows specifying the jandex index version for the build.

This is a preparation step contributing to #2204, once a jandex fix for reproducible builds is available.

Co-authored-by: Alexandre Dutra <adutra@apache.org>

snazy committed:
Built: improve reproducible archive files (#2432)

As part of the effort for #2204, this change fixes a few aspects around reproducible builds:

Some Gradle projects produce archive files, but don't get the necessary Gradle archive-tasks settings applied: one not-published project but also the tarball&zip of the distribution. This change moves the logic to the new build-plugin `polaris-reproducible`.

Another change is to have some Quarkus generated jar files adhere to the same conventions, which are constant timestamps for the zip entries and a deterministic order of the entries. That's sadly not a full fix, as the classes that are generated or instumented by Quarkus differ in each build.

Contributes to #2204

XN137 committed:
Remove commons-lang3 dependency (#2456)

outside of tests we can replace the functionality with jdk11 and guava.
also stop using `org.assertj.core.util` as its a non-public api.

jasonf20 committed:
add refresh credentials property to loadTableResult (#2341)

* add refresh credentials property to loadTableResult

* IcebergCatalogAdapterTest: Added test to ensure refresh credentials endpoint is included

* delegate refresh credential endpoint configuration to storage integration

* GCP: Add refresh credential properties

renovate-bot committed:
fix(deps): update dependency io.opentelemetry.semconv:opentelemetry-semconv to v1.37.0 (#2458)

dimas-b commented on runtime/service/src/main/java/org/apache/polaris/service/catalog/policy/CatalogPolicyServiceDefaultDelegator.java:
```diff
@@ -49,8 +54,10 @@ public Response createPolicy(
-        prefix, namespace, createPolicyRequest, realmContext, securityContext);
+    polarisEventListener.onBeforeCreatePolicy(new CatalogPolicyServiceEvents.BeforeCreatePolicyEvent(prefix, namespace, createPolicyRequest));
+    Response resp = delegate.createPolicy(prefix, namespace, createPolicyRequest, realmContext, securityContext);
+    polarisEventListener.onAfterCreatePolicy(new CatalogPolicyServiceEvents.AfterCreatePolicyEvent(prefix, namespace, resp.readEntity(LoadPolicyResponse.class)));
```
What if `delegate.createPolicy()` throws?

adnanhemani replied:
This is very possible and currently, we would just not emit a `AfterCreatePolicyEvent` (on purpose - it does not make any sense to emit an "After" event for an action that never happened).

On the mailing list, we discussed the possibility of adding `onException...` handles in the future. I am not adding that into the scope for the current instrumentation PRs. This PR does leave the door open for a future contribution that adds this. I don't personally see a need for this today, but in the future if there is one, I'm happy to contribute it at that time.

adnanhemani committed:
Add Delegator to all API Implementations (#2434)

Per the Dev ML, implements the Delegator pattern to add Events instrumentation to all Polaris APIs.

XN137 committed:
Prefer java.util.Base64 over commons-codec (#2463)

`java.util.Base64` is available since java8 and we are already using it
in a few other spots.

in a follow-up we might be able to get rid of our `commons-codec` dependency
completely.

dimas-b commented on runtime/service/src/main/java/org/apache/polaris/service/events/PolarisEventListener.java:
```diff
@@ -55,4 +55,53 @@ public void onBeforeTaskAttempted(BeforeTaskAttemptedEvent event) {}
+
+  // Catalog Policy Service Events
+  /** {@link CatalogPolicyServiceEvents.BeforeCreatePolicyEvent} */
+  public void onBeforeCreatePolicy(CatalogPolicyServiceEvents.BeforeCreatePolicyEvent event) {}
```
Why not make the `PolarisEventListener` an `interface`? I think it would be more flexible for listener implementors. All methods can have `default` implementations.

adnanhemani replied:
I'm not sure why it would become more flexible (I would like to learn what you are thinking on this topic) - but I don't see, generally, why it cannot be an interface.

If it's okay with you though, would you be okay with me making this change after all the instrumentation is added? I'd prefer to keep each PR as single-focused as possible.

flyrain committed:
Service: Move tests to the right package (#2469)

jbonofre committed:
Update versions in runtime LICENSE and NOTICE (#2468)

renovate-bot committed:
fix(deps): update dependency com.adobe.testing:s3mock-testcontainers to v4.8.0 (#2475)

renovate-bot committed:
fix(deps): update dependency com.gradleup.shadow:shadow-gradle-plugin to v9.1.0 (#2476)

adnanhemani committed:
Merge remote-tracking branch 'origin/main' into ahemani/event_instrumentation_catalog_policy_service
